homeassistant/components/xiaomi_miio/gateway.py

```diff
@@ -33,16 +42,45 @@ def gateway_info(self):
     async def async_connect_gateway(self, host, token):
         """Connect to the Xiaomi Gateway."""
         _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+
+        use_cloud = self._config_entry.options.get(CONF_CLOUD_SUBDEVICES)
+        cloud_username = self._config_entry.options.get(CONF_CLOUD_USERNAME)
+        cloud_password = self._config_entry.options.get(CONF_CLOUD_PASSWORD)
+        cloud_country = self._config_entry.options.get(CONF_CLOUD_COUNTRY)
+
         try:
             self._gateway_device = gateway.Gateway(host, token)
             # get the gateway info
             self._gateway_info = await self._hass.async_add_executor_job(
                 self._gateway_device.info
             )
+
             # get the connected sub devices
-            await self._hass.async_add_executor_job(
-                self._gateway_device.discover_devices
-            )
+            if (
+                cloud_username is not None
+                and cloud_password is not None
+                and cloud_country is not None
+                and use_cloud
+            ):
+                # use miio-cloud
+                miio_cloud = MiCloud(cloud_username, cloud_password)
+                if not await self._hass.async_add_executor_job(miio_cloud.login):
+                    _LOGGER.error(
+                        "Could not login to Xioami Miio Cloud, check the credentials"
```

Comment:
We also need to check credentials in the options flow.

Reply:
I added checks with proper errors in the options flow.